src/data/finsupp/witnesses.lean

```diff
@@ -0,0 +1,183 @@
+/-
+Copyright (c) 2022 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import data.finsupp.basic
+
+/-!
+#  Witnesses for finitely supported functions
+
+Let `α N` be two Types, and assume that `N` has a `0` and let `f g : α →₀ N` be finitely supported
+functions
+
+##  Main definitions
+
+*  `finset.diff f g : finset α`, the finite subset of `α` where `f` and `g` differ,
+*  `finset.wit f g : α`, [assuming that `a` is non-empty and `N` is linearly ordered]
+   the minimum of `finset.diff f g` (or a random element of `α` if
+   `finset.diff f g` is empty, i.e. if `f = g`).
+
+In the case in which `N` is an additive group, `finset.diff f g` coincides with
+`finset.support (f - g)`.
+-/
+
+variables {α N : Type*}
+
+namespace finsupp
+variables [decidable_eq α] [decidable_eq N]
+
+section diff
+
+section N_has_zero
+variables [has_zero N] {f g : α →₀ N}
+
+/--  Given two finitely supported functions `f g : α →₀ N`, `finsupp.diff f g` is the `finset`
+where `f` and `g` differ. -/
+def diff (f g : α →₀ N) : finset α :=
+(f.support ∪ g.support).filter (λ x, f x ≠ g x)
+
+lemma diff_comm : f.diff g = g.diff f :=
+by simp_rw [diff, finset.union_comm, ne_comm]
+
+@[simp] lemma diff_eq_empty : f.diff g = ∅ ↔ f = g :=
+begin
+  refine ⟨λ h, _, λ h, h ▸ by simp [diff]⟩,
+  ext a,
+  refine not_ne_iff.mp (λ fg, not_ne_iff.mpr h _),
+  refine finset.ne_empty_of_mem (_ : a ∈ _),
+  simp only [diff, finset.mem_filter, finset.mem_union, finsupp.mem_support_iff],
+  exact ⟨fg.ne_or_ne _, fg⟩,
+end
+
+@[simp] lemma nonempty_diff_iff : (f.diff g).nonempty ↔ f ≠ g :=
+finset.nonempty_iff_ne_empty.trans diff_eq_empty.not
+
+@[simp]
+lemma diff_zero_right : f.diff 0 = f.support :=
+by simp only [diff, coe_zero, pi.zero_apply, support_zero, finset.union_empty,
+    finset.filter_true_of_mem, mem_support_iff, imp_self, implies_true_iff]
+
+@[simp]
+lemma diff_zero_left : (0 : α →₀ N).diff f = f.support :=
+diff_comm.trans diff_zero_right
+
+@[simp]
+lemma mem_diff {a : α} : a ∈ f.diff g ↔ f a ≠ g a :=
+by simpa only [diff, finset.mem_filter, finset.mem_union, mem_support_iff, and_iff_right_iff_imp]
+    using ne.ne_or_ne _
```

Comment:
nit: I would expect this lemma (and `coe_diff`) to appear immeditately below `diff`

Reply:
Moved!